packages/common/src/api/fetchData.ts

```diff
@@ -0,0 +1,106 @@
+import axios, {AxiosError, AxiosRequestConfig, AxiosResponse} from 'axios'
+import {Api} from '@yoroi/types'
+
+type GetRequestConfig = {
+  url: string
+  method?: 'get'
+  headers?: Record<string, string>
+}
+
+type OtherRequestConfig<D = any> = {
+  url: string
+  method: 'post' | 'put' | 'delete'
+  data?: D
+  headers?: Record<string, string>
+}
+
+type RequestConfig<D = any> = GetRequestConfig | OtherRequestConfig<D>
+export type FetchData = <T, D = any>(
+  config: RequestConfig<D>,
+) => Promise<Api.Response<T>>
+
+/**
+ * Performs an HTTP request using Axios based on the specified configuration.
+ * This function simplifies making HTTP requests by handling different
+ * request methods and their respective data and headers.
+ *
+ * @param config - The configuration object for the request.
+ *                 This includes the URL, HTTP method, optional data, and headers.
+ *                 The type of `config` varies based on the HTTP method:
+ *                 - For `GET` requests, `data` should not be provided.
+ *                 - For `POST`, `PUT`, and `DELETE` requests, `data` is optional.
+ *
+ * @returns A `Promise` that resolves to the response data on a successful request
+ *          or an error object on failure. The error object includes the HTTP status
+ *          code and error message.
+ *
+ * @template T - The expected type of the response data.
+ * @template D - The type of the data to be sent with the request (for `POST`, `PUT`, `DELETE`).
+ *
+ * @example
+ * ```typescript
+ * // Example of a GET request
+ * fetchData<{ someDataType }>({
+ *   url: 'https://example.com/data',
+ * }).then(response => {
+ *   // Handle response
+ * }).catch(error => {
+ *   // Handle error
+ * })
+ * ```
+ *
+ * @example
+ * ```typescript
+ * // Example of a POST request with data
+ * fetchData<{ someDataType }, { somePayloadType }>({
+ *   url: 'https://example.com/data',
+ *   method: 'post',
+ *   data: { /* some data *\/ }
+ * }).then(response => {
+ *   // Handle response
+ * }).catch(error => {
+ *   // Handle error
+ * })
+ * ```
+ */
+export const fetchData: FetchData = <T, D = any>(
```

Comment:
legacy `fetch` was designed to throw errors, which is not great, so old `fetch` now is deprecated and whether the API errors should be throw or ignored will be a implementation decision.